### PR TITLE
Add guard against invalid block or tile size in Kamino

### PR DESCRIPTION
**Pull request by chschuma-disney**
## Description

This PR guards against some edges cases where kernel launch parameters (tile size and block size) that are computed from the problem size would be set to invalid values. Specifically, it guards against setting `block_dim=0` when there are too few constraints in a model.

Issues were detected for the kernel parameters in the ADMM solver, though just in case, a guard was also added to the conjugate solver, which could end up with `block_dim==0` if the tile size is smaller than 8 (which happens for `maxdims < 5`).

## Checklist

- [x] New or existing tests cover these changes
- [x] The documentation is up to date with these changes
- [ ] `CHANGELOG.md` has been updated (if user-facing change)

## Bug fix

**Steps to reproduce:**

1. Create a model with two bodies, each one with a sphere shape.
2. Instantiate a simulation with `SolverKamino` without any controls and contacts.
3. Run `step()` on the solver, which will run indefinitely if run on GPU.

**Minimal reproduction:**

```python
import newton
import warp as wp


def main() -> int:
    builder = newton.ModelBuilder(gravity=0.0)
    newton.solvers.SolverKamino.register_custom_attributes(builder)
    body_a = builder.add_body(
        xform=wp.transform(wp.vec3(-0.3, 0.0, 0.0), wp.quat_identity()),
        mass=1.0,
    )
    builder.add_shape_sphere(body_a, radius=0.1)
    body_b = builder.add_body(
        xform=wp.transform(wp.vec3(0.45, 0.0, 0.0), wp.quat_identity()),
        mass=1.0,
    )
    builder.add_shape_sphere(body_b, radius=0.1)
    model = builder.finalize()
    solver = newton.solvers.SolverKamino(model)

    print("Executing step...")
    state_0, state_1 = model.state(), model.state()

    solver.step(state_0, state_1, None, None, 0.01)

    return 0


if __name__ == "__main__":
    raise SystemExit(main())
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevented invalid GPU kernel launches by clamping tile/block sizes to safe minima, avoiding zero-dimension cases.
  * Improved convergence checks to use robust tile/block sizing so edge cases (e.g., minimal constraints) no longer cause invalid behavior.

* **Refactor**
  * Centralized tile/block sizing into shared helpers for consistent behavior across solvers.

* **Tests**
  * Added a unit test validating solver behavior with a single-contact configuration.

### CODE REVIEW

**Review by Guirec-Maloisel — DISMISSED** [review]
Looks good to me! I've had that case recently in the FK solver as well actually (in the unit tests where I'm checking the Jacobian for systems that might have zero actuated coordinates), but for me I was getting an exception earlier, when taking the log of zero, not sure why it didn't happen for you too.

**Comment by chschuma-disney** [comment]
> Looks good to me! I've had that case recently in the FK solver as well actually (in the unit tests where I'm checking the Jacobian for systems that might have zero actuated coordinates), but for me I was getting an exception earlier, when taking the log of zero, not sure why it didn't happen for you too.

The issue in my case was that `max_of_max_total_cts` was 3, so no error with the log, but then `tile_size // 8` becomes 0.

File: newton/_src/solvers/kamino/_src/linalg/conjugate.py [hunk 1]
```diff
@@ -580,7 +580,10 @@ def __init__(
         self.callback = callback
         self.use_cuda_graph = use_cuda_graph
 
-        self.dot_tile_size = min(2048, 2 ** math.ceil(math.log(self.maxdims, 2)))
+        if self.maxdims > 0:
+            self.dot_tile_size = min(2048, 2 ** math.ceil(math.log(self.maxdims, 2)))
+        else:
+            self.dot_tile_size = 1
         self.tiled_dot_kernel = make_dot_kernel(self.dot_tile_size, self.maxdims)
         self._allocate()
```
[the remark this answers is not in the document]
Comment (chschuma-disney) [reply newton/_src/solvers/kamino/_src/linalg/conjugate.py]: I've reverted the change. `self.maxdims > 0` is guaranteed at this point due to the check for `A.total_vec_size <= 0` earlier in the function.

File: newton/_src/solvers/kamino/_src/solvers/padmm/solver.py [hunk 2]
```diff
@@ -1198,8 +1198,11 @@ def _update_convergence_check(self, problem: DualProblem):
             problem (DualProblem): The dual forward dynamics problem to be solved.
         """
         # Compute infinity-norm of all residuals and check for convergence
-        tile_size = min(2048, 2 ** math.ceil(math.log(self._size.max_of_max_total_cts, 2)))
-        block_dim = min(256, tile_size // 8)
+        if self._size.max_of_max_total_cts > 0:
+            tile_size = min(2048, 2 ** math.ceil(math.log(self._size.max_of_max_total_cts, 2)))
+        else:
+            tile_size = 1
+        block_dim = max(1, min(256, tile_size // 8))
```
[the remark this answers is not in the document]
Comment (chschuma-disney) [reply newton/_src/solvers/kamino/_src/solvers/padmm/solver.py]: I've added a test that covers the case of two bodies with no joints and a single contact, which would result in `size.max_of_max_total_cts == 3`.

File: newton/_src/solvers/kamino/_src/linalg/conjugate.py [hunk 3]
```diff
@@ -580,7 +580,10 @@ def __init__(
         self.callback = callback
         self.use_cuda_graph = use_cuda_graph
 
-        self.dot_tile_size = min(2048, 2 ** math.ceil(math.log(self.maxdims, 2)))
+        if self.maxdims > 0:
+            self.dot_tile_size = min(2048, 2 ** math.ceil(math.log(self.maxdims, 2)))
+        else:
+            self.dot_tile_size = 1
         self.tiled_dot_kernel = make_dot_kernel(self.dot_tile_size, self.maxdims)
```
[the remark this answers is not in the document]
Comment (chschuma-disney) [reply newton/_src/solvers/kamino/_src/linalg/conjugate.py]: I've reverted the change. `self.maxdims > 0` is guaranteed at this point due to the check for `A.total_vec_size <= 0` earlier in the function.

File: newton/_src/solvers/kamino/_src/solvers/padmm/solver.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chschuma-disney) [reply newton/_src/solvers/kamino/_src/solvers/padmm/solver.py]: Since the Kamino solver is still experimental, a changelog entry is not necessary in this case.

File: newton/_src/solvers/kamino/_src/solvers/padmm/solver.py [hunk 4]
```diff
@@ -1198,8 +1198,11 @@ def _update_convergence_check(self, problem: DualProblem):
             problem (DualProblem): The dual forward dynamics problem to be solved.
         """
         # Compute infinity-norm of all residuals and check for convergence
-        tile_size = min(2048, 2 ** math.ceil(math.log(self._size.max_of_max_total_cts, 2)))
-        block_dim = min(256, tile_size // 8)
+        if self._size.max_of_max_total_cts > 0:
```
Comment (vastsoun) [inline newton/_src/solvers/kamino/_src/solvers/padmm/solver.py]: Since this `tile_size` setter op is shared between the this and on line `1245`, it might be a good idea to add a small internal helper method to return the `tile_size` and `block_dim`.

**Review by vastsoun — CHANGES_REQUESTED** [review]
Great find @chschuma-disney. Please see my minor comment.

File: newton/_src/solvers/kamino/_src/solvers/padmm/solver.py (hunk 4 above)
Comment (Guirec-Maloisel) [reply newton/_src/solvers/kamino/_src/solvers/padmm/solver.py]:
```
def get_tile_size(size: int, max_size: int = 2048):
    """Rounds up size into a power-of-two tile size, clamping to the [1, max_size] range if needed."""
    if size < 1:
        return 1
    return min(max_size, 2 ** math.ceil(math.log(size, 2)))


def get_num_tiles(size: int, tile_size: int):
    """Computes the number of tiles needed to cover an array of given size, with tiles of given size."""
    return (size + tile_size - 1) // tile_size


def get_block_size(tile_size: int, ratio: int = 8, min_size: int = 32, max_size: int = 256):
    """Computes the block size as a fixed ratio of the tile size, clamped to a min-max range."""
    return max(min_size, min(max_size, tile_size // ratio))
```

If it helps I ended up adding these to the FK solver as well as part of my open PR (taking some inspiration of this PR as well), so maybe these can be added into some utils somewhere once and for all

Comment (chschuma-disney) [reply newton/_src/solvers/kamino/_src/solvers/padmm/solver.py]: Good point. I moved these functions to their own file, and the ADMM solver is now also using them.

File: newton/_src/solvers/kamino/_src/utils/tile.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        size: Number of elements to cover.
+        tile_size: Number of elements per tile.
+
+    Returns:
+        Number of tiles needed so that ``num_tiles * tile_size >= size``.
+    """
+    return (size + tile_size - 1) // tile_size
+
+
+def get_block_size(tile_size: int, ratio: int = 8, min_size: int = 32, max_size: int = 256) -> int:
```
Comment (Guirec-Maloisel) [inline newton/_src/solvers/kamino/_src/utils/tile.py]: If it's not too much to ask, maybe it would make more sense to call this one get_block_dim() in the end, since that's how the launch argument is called?

**Review by vastsoun — APPROVED** [review]
Great work @chschuma-disney, looks great.

**Review by Guirec-Maloisel — APPROVED** [review]
Great sanitizing work, looks good to me too!
